### PR TITLE
Allowlist cli

### PR DESCRIPTION
**Pull request by pinges**
Add a script that can be used to call the allowlist smart contract funcitons

### CODE REVIEW

File: scripts/allowlist/genesisContent/README.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+			"02e472438281ece9fae629c31ebc952b0b512971efb1bacfc7d4441c586cff6c": "0000000000000000000000005b38da6a701c568545dcfcb03fcb875f56beddc4",
+			"36306db541fd1551fd93a60031e8a8c89d69ddef41d6249f5fdc265dbc8fffa2": "0000000000000000000000000000000000000000000000000000000000000101",
+			"58d9a93947083dcdedec58d43912ce0326f251a85b7701c5de5bc7d7a150676e": "0000000000000000000000000000000000000000000000000000000000000001",
+			"e0ddd3944b4f9c65a068b84248d68e1f0f838457a33299384eef528de71c9c0c": "0000000000000000000000004b20993bc481177ec7e8f571cecae8a9e22c02db",
+			"e20f19dc6931eb9e42fe3f21abe1a9ef59942d8e586871d88564d0d0b63a5e5c": "0000000000000000000000000000000000000000000000000000000000010101",
+			"0000000000000000000000000000000000000000000000000000000000000002": "0000000000000000000000000000000000000000000000000000000000000003"
+		},
+		"version": "0x01"
+	}
+The content of the file needs to be placed in the genesis file for the network. In addition the <_Address of Contract_>
```
Comment (macfarla) [inline scripts/allowlist/genesisContent/README.md:47]: nit: the words used here for the placeholders are different in the snippet above

File: scripts/allowlist/genesisContent/besu_genesis.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+		"eip150Block": 0,
+		"eip155Block": 0,
+		"eip158Block": 0,
+		"byzantiumBlock": 0,
+		"qbft": {
+			"block": 100,
+			"epochlength": 30000,
+			"blockperiodseconds": 5,
+			"requesttimeoutseconds": 10,
+			"validatorcontractaddress": "0x0000000000000000000000000000000000008888"
```
Comment (macfarla) [inline scripts/allowlist/genesisContent/besu_genesis.json:14]: nit: should we use a different default value than the permissioning smart contracts? or is this deliberate?

File: scripts/allowlist/genesisContent/genesis.json [hunk 3]
```diff
@@ -0,0 +1,53 @@
+{
```
Comment (macfarla) [inline scripts/allowlist/genesisContent/genesis.json:1]: why do we need 2 genesis files? are they different?

File: scripts/allowlist/genesisContent/package.json [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (macfarla) [inline scripts/allowlist/genesisContent/package.json:1]: did you make separate packages for a reason? would it be simpler for managing dependencies etc to combine them?

File: contracts/allowlist/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+and voting.
+
+To be able to use this contract starting from the genesis block of a blockchain the initial state of this contract 
+needs to be set in the genesis file. The `scripts/allowlist/genesisContent` directory of this 
+repository contains a javascript script that creates the storage section for this contract.
+
+The `scripts/allowlist/cli` directory contains a simple cli script that can be used to call the allowlist smart
+contract functions.
+
+### Build the Contract
```
Comment (macfarla) [inline contracts/allowlist/README.md:26]: s/Build/Compile

File: contracts/allowlist/README.md [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+* activate a validator
+* deactivate a validator
+* vote to add an account to the allowlist
+* vote to remove an account from the allowlist
+* remove votes they have cast to add or remove an account
+* execute the vote count for an account to be added or removed
+
+For a vote to be successful more than 50% of the current members of the allowlist have to vote. 
+
+To get the votes counted and a new account being added or an account to be removed the function _countVotes needs to
```
[the remark this answers is not in the document]
Comment (macfarla) [reply contracts/allowlist/README.md:13]: To get the votes counted and (if successful) the outcome enacted (the specified account added or removed), the function _countVotes needs to

[the remark this answers is not in the document]
Comment (macfarla) [reply contracts/allowlist/README.md:13]: ^ I still don't love it

File: contracts/allowlist/README.md [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+* deactivate a validator
+* vote to add an account to the allowlist
+* vote to remove an account from the allowlist
+* remove votes they have cast to add or remove an account
+* execute the vote count for an account to be added or removed
+
+For a vote to be successful more than 50% of the current members of the allowlist have to vote. 
+
+To get the votes counted and a new account being added or an account to be removed the function _countVotes needs to 
+be called.
```
Comment (macfarla) [inline contracts/allowlist/README.md:14]: do we need to mention that only one vote/election can be "open" at any one time?

**Review by macfarla — COMMENTED** [review]
left some comments!

File: contracts/allowlist/README.md [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+Accounts on the allowlist can use the API of this contract to 
+* activate a validator
+* deactivate a validator
+* vote to add an account to the allowlist
+* vote to remove an account from the allowlist
+* remove votes they have cast to add or remove an account
+* execute the vote count for an account to be added or removed
+
+For a vote to be successful more than 50% of the current members of the allowlist have to vote.
```
[the remark this answers is not in the document]
Comment (macfarla) [reply contracts/allowlist/README.md:11]: should this be election

[the remark this answers is not in the document]
Comment (macfarla) [reply contracts/allowlist/README.md:11]: For an election to be successful...

File: contracts/allowlist/README.md (hunk 7 above)
Comment (pinges) [reply contracts/allowlist/README.md:14]: That is not the case. You can vote at any time to add someone who is not on the allowlist, or remove someone who is. Votes are removed once the votes have been counted, or when the function removeVote function is called.

File: contracts/allowlist/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (pinges) [reply contracts/allowlist/README.md:11]: Done

File: contracts/allowlist/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (pinges) [reply contracts/allowlist/README.md:13]: Done

File: contracts/allowlist/ValidatorSmartContractAllowList.sol [hunk 9]
```diff
@@ -37,8 +37,9 @@ contract ValidatorSmartContractAllowList is ValidatorSmartContractInterface {
 
     address[] private validators;
     mapping(address => accountInfo) private allowedAccounts;
+    mapping(address => address) private validatorToAccount;
     uint public numAllowedAccounts;
-    mapping(address => address[]) private currentVotes; // mapping the votes for adding or removing an account to the accounts that voted for it
+    mapping(address => address[]) private currentVotes;// mapping the votes for adding or removing an account to the accounts that voted for it
 
     modifier senderIsAllowed() {
         require(allowedAccounts[msg.sender].allowed, "sender is not on the allow list");
```
[the remark this answers is not in the document]
Comment (pinges) [reply contracts/allowlist/ValidatorSmartContractAllowList.sol:45]: Done

File: scripts/allowlist/cli/allowlist.js [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+async function printEvent(eventname, receipt) {
+    const result = receipt.events[eventname].returnValues;
+    switch (eventname) {
+        case "Validator":
+            const activated = result.activated ? "activated" : "deactivated";
+            console.log(`Success: Account ${result.byAccount} has ${activated} validator ${result.validator}. Active validators: ${result.numValidators}.`);
+            break;
+        case "Vote":
+            const addRemove = result.voteToAdd ? "add" : "remove";
+            const voteRemoved = result.voteRemoved ? "removed his vote" : "has voted";
```
[the remark this answers is not in the document]
Comment (pinges) [reply scripts/allowlist/cli/allowlist.js:102]: Done

File: scripts/allowlist/genesisContent/package.json (hunk 4 above)
Comment (pinges) [reply scripts/allowlist/genesisContent/package.json:1]: Combining them is probably a good idea ...

File: README.md [hunk 11]
```diff
@@ -1,4 +1,134 @@
...
+repository contains a javascript script that creates the storage section for this contract.
+
+### Compile the Contract
+
+To create the contract code that needs to be specified in the genesis file, this contract needs to be compiled with the
+option `--bin-runtime`.
+
+After copying the interface contract into this directory
+
+    cp ../ValidatorSmartContractInterface.sol .
```
Comment (macfarla) [inline README.md:56]: does it need to be in a separate directory?

File: README.md [hunk 12]
```diff
@@ -1,4 +1,134 @@
...
+        "version": "0x01"
+    }
+
+The content of the file needs to be placed in the genesis file for the network. In addition the <_Address of Contract_>
+and <_Contract Code_> need to be filled in.
+
+An example of a genesis file using QBFT can be found in the `genesis.json` file in this directory.
+
+* The <_Address of Contract_> needs to be the same that is stated in the _qbft_ section of the genesis file for the _validatorcontractaddress_.
+* The <_Contract Code_> needs to contain the binary runtime code for the `ValidatorSmartContractAllowList.sol` contract in `contracts/allowlist`.
```
Comment (macfarla) [inline README.md:116]: is "runtime bytecode" another term for this?

File: scripts/allowlist/genesisContent/besu_genesis.json (hunk 2 above)
Comment (macfarla) [reply scripts/allowlist/genesisContent/besu_genesis.json:14]: 9999 is also used for permissioning!

Comment (macfarla) [reply scripts/allowlist/genesisContent/besu_genesis.json:14]: you can have 7777 ?

**Review by macfarla — APPROVED** [review]
looking good! a few nitpicks

File: README.md (hunk 12 above)
Comment (pinges) [reply README.md:116]: Not sure whether that is the same ...

File: README.md (hunk 11 above)
Comment (pinges) [reply README.md:56]: The problem is that you need to allow solidity to access files other than the current directory and it's subdirectories. That makes the command line quite ugly ....

File: README.md [hunk 13]
```diff
@@ -1,4 +1,134 @@
...
+
+**See the web3-js based script in _scripts/allowlist/cli_ for a simple cli script to call the allowlist smart contract functions.**
+
+Accounts on the allowlist can use the API of this contract to
+* activate a validator
+* deactivate a validator
+* vote to add an account to the allowlist
+* vote to remove an account from the allowlist
+* remove votes they have cast to add or remove an account
+* execute the vote count for an account to be added or removed
```
[the remark this answers is not in the document]
Comment (pinges) [reply README.md:35]: Haven't changed that one ...

File: manualTest/README.md [hunk 14]
```diff
@@ -0,0 +1,49 @@
+# Manual Testing
+
+This directory and it's subdirectories contain configuration files and other files that can be used to create a
```
Comment (macfarla) [inline manualTest/README.md:3]: s/it's/its

File: manualTest/README.md [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+    besu --genesis-file qbftNode2/besu_genesis.json --data-path qbftNode2/data --node-private-key-file=qbftNode2/val_key_1 --rpc-http-enabled --rpc-http-apis=ETH,NET,DEBUG,ADMIN,WEB3,EEA,PRIV,QBFT,TRACE --rpc-http-port=8591 --min-gas-price=0 --p2p-port=30305 --network-id=2017 --host-allowlist="all" --rpc-http-cors-origins="all"
+
+
+Validators:
+ * Besu Node 1 account: 0x9A6d82Ef3912d5aB60473124BCCd2f2A640769D7, private key: 0x70f1384b24df3d2cdaca7974552ec28f055812ca5e4da7a0ccd0ac0f8a4a9b00
+ * Besu Node 2 account: 0x65463BF6268e5cC409b6501eC846487B935A1446, private key: 0xad0352cfc09aa0128db4e135fcea276523c400163dcc762a11ecba29d5f0a34a
+
+Allowed Accounts:
+ * 0xfe3b557e8fb62b89f4916b721be55ceb828dbd73, private key: 0x8f2a55949038a9610f50fb23b5883af3b4ecb3c3bb792cbcefbd1542c692be63 
+   (acitve validator in genesis block is 0x9A6d82Ef3912d5aB60473124BCCd2f2A640769D7)
```
Comment (macfarla) [inline manualTest/README.md:23]: s/acitve/active

File: manualTest/README.md [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+ * 0x627306090abaB3A6e1400e9345bC60c78a8BEf57, private key: 0xc87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3
+   (acitve validator in genesis block is 0x65463BF6268e5cC409b6501eC846487B935A1446)
```
Comment (macfarla) [inline manualTest/README.md:25]: s/acitve/active

File: manualTest/README.md [hunk 17]
```diff
@@ -0,0 +1,49 @@
+# Manual Testing
```
Comment (macfarla) [inline manualTest/README.md:1]: I'd maybe rename the manualTest dir demoSetup or sampleSetup

File: manualTest/README.md (hunk 14 above)
Comment (pinges) [reply manualTest/README.md:3]: Done

File: manualTest/README.md (hunk 16 above)
Comment (pinges) [reply manualTest/README.md:25]: Done

File: manualTest/README.md (hunk 17 above)
Comment (pinges) [reply manualTest/README.md:1]: Done

**Pull request merged**